Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+With this, the earlier code snippet can be replaced with something much easier to read, maintain, and debug:
+
+```swift
+/// Instantiate services
+let postgresClient = PostgresClient()
+
+/// Instantiate LambdaRuntime with a closure handler implementing the business logic of the Lambda function
+let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+    /// Use initialized service within the handler
+    try await postgresClient.query(...)
+}
+
+/// Use ServiceLifecycle to manage the initialization and termination
+/// of the services as well as the LambdaRuntime
+let serviceGroup = ServiceGroup(
+    services: [postgresClient, runtime],
+    configuration: .init(gracefulShutdownSignals: [.sigterm]),
+    logger: logger
+)
+try await serviceGroup.run()
+```
+
+#### Simplifying Codable support
+
+A detailed explanation is provided in the Codable Support section. In short, much of the boilerplate code defined for each handler protocol in `Lambda+Codable` and `Lambda+String` will be replaced with a single `LambdaCodableAdapter` class.
+
+This adapter class is generic over (1) any handler conforming to `LambdaHandler`, (2) the user-specified input and output types, and (3) any decoder and encoder conforming to `LambdaEventDecoder` and `LambdaOutputDecoder`. The adapter will wrap the underlying handler with encoding/decoding logic.
+
+#### Simplifying the handler protocols
+
+There are four different handler protocols in the current API (`SimpleLambdaHandler`, `LambdaHandler`, `EventLoopLambdaHandler`, `ByteBufferLambdaHandler`). As noted in the **Current Limitations** section, the ease-of-use varies for each handler protocol and users may not be able to easily determine which protocol best serves their use-case without spending time digging into the library. To reduce this problem and provide users with clear-cut options, we propose replacing all of the existing handler protocols with just two: `LambdaHandler` and `StreamingLambdaHandler`. Both will be explained in the **Detailed Solution** section.
+
+## Detailed Solution
+
+Below you’ll find explanation’s for all types that we want to use in AWS Lambda Runtime v2.
+
+### LambdaResponseWriter
+
+We will introduce a new `LambdaResponseWriter` protocol. It is used in the new `StreamingLambdaHandler` (defined below), which is the new base protocol for the `LambdaRuntime` (defined below as well).
+
+```swift
+/// A writer object to write the Lambda response stream into
+public protocol LambdaResponseWriter: ~Copyable {
+    /// Write a response part into the stream
+    func write(_ buffer: ByteBuffer) async throws
+    /// End the response stream
+    consuming func finish()
+    /// Write a response part into the stream and end the response stream
+    consuming func writeAndFinish(_ buffer: ByteBuffer) async throws
+    /// Report an error in the response stream
+    consuming func reportError(_ error: any Error) async throws
```

Comment:
This seems like it is finishing as well? Do just want to spell this `finish(error: (any Error)?)` instead?

Reply:
Not sure I agree here. I think the `reportError` is more explicit in what it does, than a finish with an error.

Reply:
This seems to dovetail with the comments about making handle support throwing of errors. Do we want users to report errors, via `reportError` or via throwing an error. Given this will end the stream I think I agree with Franz.

Reply:
After thinking about this more, I think we should just make the handler method throwing. In this case we will do the error reporting, which means users don't have to deal with it, which in turn means we should remove this method here.

@FranzBusch, @tachyonics, @adam-fowler Can you get behind this?

Reply:
The only issue with that is the user can't customise their response to the error, they might want to report the error differently if they haven't started streaming response data versus being in the middle of streaming a response.

Reply:
They can still do this by doing a custom do/catch. It just makes the default case easier